### PR TITLE
feat(amp): added prometheusservice rulegroupsnamespace

### PR DESCRIPTION
**Pull request by r0mdau**
### Description of your changes

Added prometheusservice/amp RuleGroupsNamespace resource managing Prometheus recording rules.

Using code generator.

Fixes half of #1341

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

```
$ k apply -f examples/prometheusservice/workspace.yaml 
workspace.prometheusservice.aws.crossplane.io/example created

$ k get workspace
NAME      READY   SYNCED   EXTERNAL-NAME
example   True    True     ws-766f58a6-5b09-4f53-b9d9-9a7cba11a7cd

$ k apply -f examples/prometheusservice/rulegroupsnamespace.yaml 
rulegroupsnamespace.prometheusservice.aws.crossplane.io/example-rule created
$ k get rulegroupsnamespace
NAME           READY   SYNCED   EXTERNAL-NAME
example-rule   True    True     arn:aws:aps:us-east-1:015890463774:rulegroupsnamespace/ws-766f58a6-5b09-4f53-b9d9-9a7cba11a7cd/example-rule
```

```
$ k describe rulegroupsnamespace example-rule 
Name:         example-rule
Namespace:    
Labels:       <none>
Annotations:  crossplane.io/external-create-pending: 2022-07-19T22:34:53-04:00
              crossplane.io/external-create-succeeded: 2022-07-19T22:34:54-04:00
              crossplane.io/external-name: arn:aws:aps:us-east-1:015890463774:rulegroupsnamespace/ws-766f58a6-5b09-4f53-b9d9-9a7cba11a7cd/example-rule
API Version:  prometheusservice.aws.crossplane.io/v1alpha1
Kind:         RuleGroupsNamespace
Metadata:
  Creation Timestamp:  2022-07-20T02:34:53Z
  Finalizers:
    finalizer.managedresource.crossplane.io
  Generation:  2
  Managed Fields:
    API Version:  prometheusservice.aws.crossplane.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:deletionPolicy:
        f:forProvider:
          .:
          f:data:
          f:name:
          f:region:
          f:workspaceID:
        f:providerConfigRef:
          .:
          f:name:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2022-07-20T02:34:53Z
    API Version:  prometheusservice.aws.crossplane.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:crossplane.io/external-create-pending:
          f:crossplane.io/external-create-succeeded:
          f:crossplane.io/external-name:
        f:finalizers:
          .:
          v:"finalizer.managedresource.crossplane.io":
      f:spec:
        f:forProvider:
          f:tags:
            .:
            f:crossplane-kind:
            f:crossplane-name:
            f:crossplane-providerconfig:
      f:status:
        .:
        f:atProvider:
          .:
          f:arn:
          f:status:
            .:
            f:statusCode:
            f:statusReason:
        f:conditions:
    Manager:         main
    Operation:       Update
    Time:            2022-07-20T02:34:54Z
  Resource Version:  1018540
  UID:               ae61bf67-c54e-453a-8725-06818fc6cfa9
Spec:
  Deletion Policy:  Delete
  For Provider:
    Data:    Z3JvdXBzOgogIC0gbmFtZTogdGVzdAogICAgcnVsZXM6CiAgICAtIHJlY29yZDogbWV0cmljOnJlY29yZGluZ19ydWxlCiAgICAgIGV4cHI6IGF2ZyhyYXRlKGNvbnRhaW5lcl9jcHVfdXNhZ2Vfc2Vjb25kc190b3RhbFs1bV0pKQogIC0gbmFtZTogYWxlcnQtdGVzdAogICAgcnVsZXM6CiAgICAtIGFsZXJ0OiBtZXRyaWM6YWxlcnRpbmdfcnVsZQogICAgICBleHByOiBhdmcocmF0ZShjb250YWluZXJfY3B1X3VzYWdlX3NlY29uZHNfdG90YWxbNW1dKSkgPiAwCiAgICAgIGZvcjogNW0=
    Name:    example-rule
    Region:  us-east-1
    Tags:
      Crossplane - Kind:            rulegroupsnamespace.prometheusservice.aws.crossplane.io
      Crossplane - Name:            example-rule
      Crossplane - Providerconfig:  default
    Workspace ID:                   ws-766f58a6-5b09-4f53-b9d9-9a7cba11a7cd
  Provider Config Ref:
    Name:  default
Status:
  At Provider:
    Arn:  arn:aws:aps:us-east-1:015890463774:rulegroupsnamespace/ws-766f58a6-5b09-4f53-b9d9-9a7cba11a7cd/example-rule
    Status:
      Status Code:    ACTIVE
      Status Reason:  
  Conditions:
    Last Transition Time:  2022-07-20T02:36:55Z
    Reason:                Available
    Status:                True
    Type:                  Ready
    Last Transition Time:  2022-07-20T02:34:54Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced
Events:
  Type     Reason                           Age    From                                                             Message
  ----     ------                           ----   ----                                                             -------
  Normal   CreatedExternalResource          4m33s  managed/rulegroupsnamespace.prometheusservice.aws.crossplane.io  Successfully requested creation of external resource
```

rulegroupsnamespace

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by haarchri** [comment]
@r0mdau  can you check changed chmod stuff and new files for *.sh ? - please cleanup the PR - in general i tested the PR LGTM

**Comment by r0mdau** [comment]
Nice catch @haarchri , I verified files and all are in sync now with permissions.

**Comment by haarchri** [comment]
Think it's better to get data out of a SecretRef ? So it's easy to create a secret out of a file and use it in the resource - what do you think ?

**Comment by r0mdau** [comment]
I looked into other apis in this provider and I only see the use of k8s Secrets for credentials / passwords which are secrets.

On our side we will not create alerting and recording rules by hand in the k8s API. We will always have an application or at least a bash command in a Gitlab runners able to transform the original file in a base64 encoded string.

Creating them by hand will occur but only for testing or debugging. We also plan to create hundreds of them, so it is not the same to deal with only 1 CR or 1 CR and 1 secret for each in etcd.

As admin, reading a secret with `kubectl` will still show a base64 encoded string, no gain on that.

We may need comments from other.

File: examples/prometheusservice/rulegroupsnamespace.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+apiVersion: prometheusservice.aws.crossplane.io/v1alpha1
+kind: RuleGroupsNamespace
+metadata:
+  name: example-rule
+spec:
+  forProvider:
+    region: us-east-1
+    name: example-rule
+    workspaceID: ws-00000000-0000-0000-0000-000000000000
```
Comment (haarchri) [inline examples/prometheusservice/rulegroupsnamespace.yaml:9]: can we use here workspaceIdRef?

File: apis/prometheusservice/v1alpha1/custom_types.go [hunk 2]
```diff
@@ -16,5 +16,28 @@ limitations under the License.
...
 // CustomWorkspaceParameters includes custom fields about WorkspaceParameters.
 type CustomWorkspaceParameters struct{}
+
+// CustomRuleGroupsNamespaceParameters includes custom fields about CustomRuleGroupsNamespaceParameters.
+// workspaceID is actually required but since it's reference-able, it's not marked as required.
+type CustomRuleGroupsNamespaceParameters struct {
+	// workspaceID is the ID for the Workspace.
+	// +immutable
+	// +crossplane:generate:reference:type=Workspace
+	WorkspaceID *string `json:"workspaceID,omitempty"`
```
Comment (haarchri) [inline apis/prometheusservice/v1alpha1/custom_types.go:32]: `json:"workspaceId...

File: apis/prometheusservice/v1alpha1/custom_types.go [hunk 3]
```diff
@@ -16,5 +16,28 @@ limitations under the License.
...
+
+	// WorkspaceIDRef is a reference to a Workspace used to set
+	// the WorksworkspaceIDpaceId.
+	// +optional
+	WorkspaceIDRef *xpv1.Reference `json:"workspaceIDRef,omitempty"`
```
Comment (haarchri) [inline apis/prometheusservice/v1alpha1/custom_types.go:37]: `json:"workspaceIdRef...

File: apis/prometheusservice/v1alpha1/custom_types.go [hunk 4]
```diff
@@ -16,5 +16,28 @@ limitations under the License.
...
+
+	// WorkspaceIDSelector selects references to Workspace used
+	// to set the workspaceID.
+	// +optional
+	WorkspaceIDSelector *xpv1.Selector `json:"workspaceIDSelector,omitempty"`
```
Comment (haarchri) [inline apis/prometheusservice/v1alpha1/custom_types.go:42]: `json:"workspaceIdSelector

**Review by haarchri — CHANGES_REQUESTED** [review]
few nitpicks - after that we ready to merge

**Comment by haarchri** [comment]
@r0mdau checked again our Implementation in jet-provider which is also without Secret - so we can go Forward - thanks

**Comment by r0mdau** [comment]
> few nitpicks - after that we ready to merge

I use `workspaceID` because of : https://github.com/crossplane-contrib/provider-aws/blob/master/apis/prometheusservice/v1alpha1/zz_types.go#L70

Are you sure I operate this change ?

File: examples/prometheusservice/rulegroupsnamespace.yaml (hunk 1 above)
Comment (r0mdau) [reply examples/prometheusservice/rulegroupsnamespace.yaml:9]: Yes, I am waiting your answer here https://github.com/crossplane-contrib/provider-aws/pull/1390#issuecomment-1197440184 And then I will use workspace id ref

**Comment by haarchri** [comment]
Because of Jet Provider i prefer Id better for bidirectional migration

**Review by haarchri — APPROVED** [review]
Thanks for Implementation

**Pull request merged**